client/src/main/java/io/pravega/client/segment/impl/SegmentInputStreamImpl.java

```diff
@@ -194,12 +204,29 @@ private void verifyIsAtCorrectOffset(WireCommands.SegmentRead segmentRead) {
     }
 
     /**
-     * Issues a request if there is enough room for another request, and we aren't already waiting on one
+     * Issues a request
+     *  - if there is enough room for another request, and we aren't already waiting on one and
+     *  - if we have not read up to the configured endOffset.
      */
     private void issueRequestIfNeeded() {
-        if (!receivedEndOfSegment && !receivedTruncated && buffer.capacityAvailable() >= readLength && outstandingRequest == null) {
-            outstandingRequest = asyncInput.read(offset + buffer.dataAvailable(), readLength);
+        //compute read length based on current offset upto which the events are read are read.
+        int updatedReadLength = computeReadLength(offset + buffer.dataAvailable(), readLength);
+        if (!receivedEndOfSegment && !receivedTruncated && updatedReadLength > 0 && buffer.capacityAvailable() >= updatedReadLength && outstandingRequest == null) {
+            outstandingRequest = asyncInput.read(offset + buffer.dataAvailable(), updatedReadLength);
+        }
+    }
+
+    /*
+     * Compute the read length based on the current fetch offset and the configured end offset.
+     */
+    private int computeReadLength(long currentFetchOffset, int currentReadLength) {
+        Preconditions.checkState(endOffset >= currentFetchOffset,
+                "Current offset up to to which events are fetched should be less than the configured end offset");
+        if (Long.MAX_VALUE == endOffset) { //endOffset is Long.MAX_VALUE if the endOffset is not set.
+            return currentReadLength;
         }
+        long numberOfBytesToRead = endOffset - currentFetchOffset;
+        return Math.min(currentReadLength, Math.toIntExact(numberOfBytesToRead));
```

Comment:
this line will throw if endOffset is say 5GB and currentFetchOffset is 0.

Reply:
corrected it.